udp_driver/src/udp_component.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "udp_component/udp_component.hpp"
+
+#include <vector>
+
+#include "rclcpp/rclcpp.hpp"
+
+
+namespace udp_component
+{
+Packet::Packet(std::vector<uint8_t> data)
+: data{data} {}
+Packet::Packet()
+: data{0} {}
+
+UdpComponent::UdpComponent(const rclcpp::NodeOptions & options)
+: UdpDriverT("udp_component", options)
+{
+  RCLCPP_INFO(this->get_logger(), "Initializing udp_component");
+  while (rclcpp::ok()) {
+    run(100U);
+  }
+}
```

Comment:
So, the `run()` function blocks continuously while `rclcpp::ok()` internally. Putting it in another `while()` loop isn't necessary nor is providing a number of max iterations. Here is what I would recommend:

```suggestion
{
  RCLCPP_INFO(this->get_logger(), "Initializing udp_component");
  std::thread reader_thread(&UdpComponent::run(), this);
}
```

Then, in a destructor, check if the thread is join-able and, if so, join it. The `run()` function is pretty well-behaved so putting the join in the destructor should be fine.

This way the node isn't 100% tied up with just receiving data. This will especially be important once we implement sending.

Reply:
so I tried it this way but it threw a compile error complaining something about the arguments of the `run` function:

```
error: static assertion failed: std::thread arguments must be invocable after conversion to rvalues
```

i just committed an update where im passing in an additional `0U` argument and it is compiling but I'm not sure if this is the right way to do that? ill test it tomorrow on a udp device.